File: speech_to_text/csharp/real_time_transcription/Realtime/Realtime.cs
using System;
using System.IO;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

class Realtime
{
    static async Task Main()
    {
        var apiKey = Environment.GetEnvironmentVariable("SONIOX_API_KEY");

        var wsUri = new Uri("wss://stt-rt.soniox.com/transcribe-websocket");

        using (ClientWebSocket ws = new ClientWebSocket())
        {
            // Connect to WebSocket API
            Console.WriteLine("Opening WebSocket connection...");
            await ws.ConnectAsync(wsUri, CancellationToken.None);

            // Send start request with correct field names
            var startMessage = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(new
            {
                api_key = apiKey,
                audio_format = "pcm_s16le",
                sample_rate = 16000,
                num_channels = 1,
                model = "stt-rt-preview",
                language_hints = Array.Empty<string>()
            }));
            await ws.SendAsync(new ArraySegment<byte>(startMessage), WebSocketMessageType.Text, true, CancellationToken.None);

            // Start send audio task
            var sendAudioTask = Task.Run(async () =>
            {
                using (FileStream fs = new FileStream("coffee_shop.pcm_s16le", FileMode.Open, FileAccess.Read))
                {
                    Console.WriteLine("Transcription started.");
                    var buffer = new byte[3840];
                    int bytesRead;
                    while ((bytesRead = await fs.ReadAsync(buffer, 0, buffer.Length)) > 0)
                    {
                        if (ws.State != WebSocketState.Open)
                        {
                            return;
                        }

                        var audioChunk = new ArraySegment<byte>(buffer, 0, bytesRead);
                        await ws.SendAsync(audioChunk, WebSocketMessageType.Binary, true, CancellationToken.None);

                        await Task.Delay(120);
                    }

                    // Send end of file (empty message)
                    await ws.SendAsync(new ArraySegment<byte>(Array.Empty<byte>()), WebSocketMessageType.Text, true, CancellationToken.None);
                }
            });

            try
            {
                // Receive text
                var buffer = new byte[8192];

                while (true)
                {
                    if (ws.State != WebSocketState.Open)
                    {
                        break;
                    }

                    var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);

                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
                        break;
                    }

                    if (result.MessageType == WebSocketMessageType.Text)
                    {
                        string message = Encoding.UTF8.GetString(buffer, 0, result.Count);
                        var res = JsonSerializer.Deserialize<JsonElement>(message);

                        // Handle any possible transcription error
                        if (res.TryGetProperty("error_code", out var errorCode))
                        {
                            res.TryGetProperty("error_message", out var errorMessage);
                            Console.WriteLine("");
                            Console.WriteLine("Error {0}: {1}", errorCode.GetInt32(), errorMessage.GetString());
                            await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
                            break;
                        }

                        // Process tokens
                        if (res.TryGetProperty("tokens", out var tokens))
                        {
                            foreach (var token in tokens.EnumerateArray())
                            {
                                if (token.TryGetProperty("text", out var text))
                                {
                                    Console.Write(text.GetString());
                                    Console.Out.Flush();
                                }
                            }
                        }

                        if (res.TryGetProperty("finished", out var finished) && finished.GetBoolean())
                        {
                            // Close the WebSocket
                            Console.WriteLine("\nTranscription done.");
                            await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
                            break;
                        }
                    }
                }

                Console.WriteLine();
            }
            finally
            {
                await sendAudioTask;
            }
        }
    }
}

File: speech_to_text/python/real_time_transcription/realtime/realtime.py
import json
import os
import threading
import time
from websockets import ConnectionClosedOK, ConnectionClosedError
from websockets.sync.client import connect

WEBSOCKET_URL = "wss://stt-rt.soniox.com/transcribe-websocket"
FILE_TO_STREAM = "coffee_shop.pcm_s16le"

# Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
API_KEY = os.environ.get("SONIOX_API_KEY")
if API_KEY is None:
    raise KeyError(
        "Your API key is missing. Please set the SONIOX_API_KEY environment variable."
    )

# Connect to WebSocket API
print("Opening WebSocket connection...")
with connect(WEBSOCKET_URL) as ws:
    # Send start request
    ws.send(
        json.dumps(
            {
                "api_key": API_KEY,
                "audio_format": "pcm_s16le",
                "sample_rate": 16000,
                "num_channels": 1,
                "model": "stt-rt-preview",
                "language_hints": [],
            }
        )
    )

    def send_audio():
        # Read and send audio data from file over WebSocket connection
        with open(FILE_TO_STREAM, "rb") as fh:
            while True:
                data = fh.read(3840)
                if len(data) == 0:
                    break
                ws.send(data)
                # Sleep for 120 ms
                time.sleep(0.12)

        # Signal end of file
        ws.send("")

    # Start a thread to send audio
    send_audio_thread = threading.Thread(target=send_audio)
    send_audio_thread.start()
    
    print("Transcription started.")

    # Receive and process text messages
    try:
        while True:
            message = ws.recv()
            res = json.loads(message)

            if res.get("error_code"):
                print(f'\nError: {res["error_code"]} {res["error_message"]}')
                break

            for token in res.get("tokens", []):
                if token.get("text"):
                    # Print out transcribed words
                    print(token["text"], end="", flush=True)
            
            if res.get("finished"):
                print("\nTranscription done.")

    except ConnectionClosedOK:
        pass

    except ConnectionClosedError as ex:
        print("\nConnection error occurred: ", ex)

    finally:
        send_audio_thread.join()
